### PR TITLE
statistics: using standard lib slices instead of exp slices

### LINKED ISSUE #45933: Migration to Standard Library for exp/slices and exp/maps Packages
## Enhancement

With the latest release of Go, we have introduced the new slices and maps packages in the standard library. These packages provide improved functionality and support generic operations on slices and maps of any element or key type.
To leverage the benefits of the standard library and ensure better code maintainability, we need to migrate our codebase from the exp/slices and exp/maps packages to the standard library equivalents.

### PR DESCRIPTION
**Pull request by fatelei**
### What problem does this PR solve?

using builtin slices lib instead of exp slices 

Issue Number: ref #45933

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by hawkingrei** [comment]
@fatelei Thank you.

TiDB use bazel to speed up build/test/linter. so you need to run ```make bazel_prepare``` to update the bazel config.

**Comment by hawkingrei** [comment]
> @fatelei Thank you.
> 
> TiDB use bazel to speed up build/test/linter. so you need to run `make bazel_prepare` to update the bazel config.

CI will tell you where to change. you can update config by it.

**Review by hawkingrei — APPROVED** [review]

File: statistics/cmsketch.go [hunk 1]
```diff
@@ -919,11 +920,11 @@ func checkEmptyTopNs(topNs []*TopN) bool {
 
 // SortTopnMeta sort topnMeta
 func SortTopnMeta(topnMetas []TopNMeta) []TopNMeta {
-	slices.SortFunc(topnMetas, func(i, j TopNMeta) bool {
+	slices.SortFunc(topnMetas, func(i, j TopNMeta) int {
 		if i.Count != j.Count {
-			return i.Count > j.Count
+			return cmp.Compare(i.Count, j.Count)
```
Comment (hi-rustin) [inline statistics/cmsketch.go:925]: I think this is the opposite of the original logic. Also, could you please add some tests to cover this?

**Review by hi-rustin — CHANGES_REQUESTED** [review]

File: statistics/cmsketch.go (hunk 1 above)
Comment (fatelei) [reply statistics/cmsketch.go:925]: yes，need desc order

File: statistics/cmsketch_test.go [hunk 2]
```diff
@@ -16,6 +16,7 @@ package statistics
 
 import (
 	"fmt"
+	"github.com/stretchr/testify/assert"
```
Comment (hawkingrei) [inline statistics/cmsketch_test.go:19]: You need to sort the import.

**Comment by fatelei** [comment]
> bazel_prepare

done

**Comment by hawkingrei** [comment]
@fatelei Please fix the linter.


```
 statistics/analyze.go:1:1: 
--- statistics/cmsketch_test.go
+++ statistics/cmsketch_test.go
@@ -21,14 +21,13 @@
 	"testing"
 	"time"
 
-	"github.com/stretchr/testify/assert"
-
 	"github.com/pingcap/errors"
 	"github.com/pingcap/tidb/parser/mysql"
 	"github.com/pingcap/tidb/sessionctx/stmtctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/codec"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 ) 
```

**Comment by fatelei** [comment]
> @fatelei Please fix the linter.
> 
> ```
>  statistics/analyze.go:1:1: 
> --- statistics/cmsketch_test.go
> +++ statistics/cmsketch_test.go
> @@ -21,14 +21,13 @@
>  	"testing"
>  	"time"
>  
> -	"github.com/stretchr/testify/assert"
> -
>  	"github.com/pingcap/errors"
>  	"github.com/pingcap/tidb/parser/mysql"
>  	"github.com/pingcap/tidb/sessionctx/stmtctx"
>  	"github.com/pingcap/tidb/types"
>  	"github.com/pingcap/tidb/util/chunk"
>  	"github.com/pingcap/tidb/util/codec"
> +	"github.com/stretchr/testify/assert"
>  	"github.com/stretchr/testify/require"
>  ) 
> ```

done

**Comment by hawkingrei** [comment]
@fatelei Some tests have failed, please check them.

File: statistics/cmsketch.go [hunk 3]
```diff
@@ -934,11 +935,11 @@ func GetMergedTopNFromSortedSlice(sorted []TopNMeta, n uint32) (*TopN, []TopNMet
 }
 
 func getMergedTopNFromSortedSlice(sorted []TopNMeta, n uint32) (*TopN, []TopNMeta) {
-	slices.SortFunc(sorted, func(i, j TopNMeta) bool {
+	slices.SortFunc(sorted, func(i, j TopNMeta) int {
 		if i.Count != j.Count {
-			return i.Count > j.Count
+			return cmp.Compare(i.Count, j.Count)
```
Comment (hawkingrei) [inline statistics/cmsketch.go:940]:
```suggestion
			return cmp.Compare(j.Count, i.Count)
```

File: statistics/cmsketch_test.go [hunk 4]
```diff
@@ -339,6 +340,18 @@ func TestMergePartTopN2GlobalTopNWithoutHists(t *testing.T) {
...
+func TestSortTopnMeta(t *testing.T) {
+	data := []TopNMeta{{
+		Encoded: []byte("a"),
+		Count:   1,
+	}, {
+		Encoded: []byte("b"),
+		Count:   2,
+	}}
+	sortedData := SortTopnMeta(data)
+	assert.Equal(t, uint64(2), sortedData[0].Count)
```
Comment (hawkingrei) [inline statistics/cmsketch_test.go:352]:
```suggestion
	require.Equal(t, uint64(2), sortedData[0].Count)
```

**Comment by hawkingrei** [comment]
@fatelei Please run ```make bazel_preapre``` to update bazel config.

```
index 5c2461120..bd200e961 100644

--- a/statistics/BUILD.bazel

+++ b/statistics/BUILD.bazel

@@ -120,7 +120,6 @@ go_test(

         "@com_github_pingcap_errors//:errors",

         "@com_github_pingcap_failpoint//:failpoint",

         "@com_github_pingcap_log//:log",

-        "@com_github_stretchr_testify//assert",

         "@com_github_stretchr_testify//require",

         "@org_uber_go_goleak//:goleak",

         "@org_uber_go_zap//:zap",

```

**Comment by fatelei** [comment]
> @fatelei Please run `make bazel_preapre` to update bazel config.
> 
> ```
> index 5c2461120..bd200e961 100644
> 
> --- a/statistics/BUILD.bazel
> 
> +++ b/statistics/BUILD.bazel
> 
> @@ -120,7 +120,6 @@ go_test(
> 
>          "@com_github_pingcap_errors//:errors",
> 
>          "@com_github_pingcap_failpoint//:failpoint",
> 
>          "@com_github_pingcap_log//:log",
> 
> -        "@com_github_stretchr_testify//assert",
> 
>          "@com_github_stretchr_testify//require",
> 
>          "@org_uber_go_goleak//:goleak",
> 
>          "@org_uber_go_zap//:zap",
> ```

done

**Review by hawkingrei — APPROVED** [review]

File: statistics/cmsketch.go [hunk 5]
```diff
@@ -919,11 +920,11 @@ func checkEmptyTopNs(topNs []*TopN) bool {
 
 // SortTopnMeta sort topnMeta
 func SortTopnMeta(topnMetas []TopNMeta) []TopNMeta {
-	slices.SortFunc(topnMetas, func(i, j TopNMeta) bool {
+	slices.SortFunc(topnMetas, func(i, j TopNMeta) int {
 		if i.Count != j.Count {
-			return i.Count > j.Count
+			return cmp.Compare(j.Count, i.Count)
```
Comment (hi-rustin) [inline statistics/cmsketch.go:925]: Could please add a note here?

File: statistics/cmsketch_test.go [hunk 6]
```diff
@@ -339,6 +339,18 @@ func TestMergePartTopN2GlobalTopNWithoutHists(t *testing.T) {
 	require.Len(t, leftTopN, 1, "should have 1 left topN")
 }
 
+func TestSortTopnMeta(t *testing.T) {
```
Comment (hi-rustin) [inline statistics/cmsketch_test.go:342]: 👍🏿

**Review by hi-rustin — APPROVED** [review]
Thanks!

**Pull request merged**